File: 1/2021.08.25/797.All Paths From Source to Target/python/solution.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        # DFS
        def DFS(x, tmp):
            # terminator
            if x == len(graph) - 1:
                res.append(tmp[:])
                return
            for y in graph[x]:
                tmp.append(y)
                DFS(y, tmp)
                tmp.pop()
        
        res = []
        DFS(0, [0])
        return res

File: 1/2021.08.25/797.All Paths From Source to Target/java/solution.java
class Solution {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        // BFS
        List<List<Integer>> res = new ArrayList<>();
        Set<List<Integer>> level = new HashSet<>();
        level.add(new ArrayList<>(Arrays.asList(0)));
        while (!level.isEmpty()) {
            Set<List<Integer>> nextLevel = new HashSet<>();
            for (List<Integer> path : level) {
                for (int node : graph[path.get(path.size() - 1)]) {
                    List<Integer> newPath = new ArrayList<>(path);
                    newPath.add(node);
                    if (node == graph.length - 1) {
                        res.add(newPath);
                    } else {
                        nextLevel.add(newPath);
                    }
                }
            }
            level = nextLevel;
        }
        return res;
    }
}
